client/web/src/components/widgets/options/ToolOptions.vue

```diff
@@ -0,0 +1,131 @@
+<template>
+	<div class="tool-options">
+		<template v-for="(option, index) in optionsMap.get(activeTool) || []" :key="index">
+			<IconButton v-if="option.kind === 'icon'" :icon="option.icon" :size="24" :title="option.title" />
+			<Separator v-if="option.kind === 'separator'" :type="option.type" />
+			<PopoverButton v-if="option.kind === 'popover'">
+				<h3>{{ option.title }}</h3>
+				<p>{{ option.placeholder_text }}</p>
+			</PopoverButton>
+			<NumberInput v-if="option.kind === 'number'" :callback="option.callback" :initialValue="option.initial" :step="option.step" :min="option.min" :updateOnCallback="true" />
+		</template>
+	</div>
+</template>
+
+<style lang="scss">
+.tool-options {
+	height: 100%;
+	flex: 0 0 auto;
+	display: flex;
+	align-items: center;
+}
+</style>
+
+<script lang="ts">
+import { defineComponent } from "vue";
+import Separator, { SeparatorType } from "@/components/widgets/separators/Separator.vue";
+import IconButton from "@/components/widgets/buttons/IconButton.vue";
+import PopoverButton from "@/components/widgets/buttons/PopoverButton.vue";
+import NumberInput from "@/components/widgets/inputs/NumberInput.vue";
+
+const wasm = import("@/../wasm/pkg");
+
+type ToolOptionsList = Array<ToolOptions>;
+type ToolOptionsMap = Map<string, ToolOptionsList>;
+
+type ToolOptions = IconOption | SeparatorOption | PopoverOption | NumberOption;
+
+interface IconOption {
+	kind: "icon";
+	icon: string;
+	title: string;
+}
+
+interface SeparatorOption {
+	kind: "separator";
+	type: SeparatorType;
+}
+
+interface PopoverOption {
+	kind: "popover";
+	title: string;
+	placeholder_text: string;
+}
+
+interface NumberOption {
+	kind: "number";
+	initial: number;
+	step: number;
+	min?: number;
+	callback?: Function;
+}
+
+export default defineComponent({
+	props: {
+		activeTool: { type: String },
+	},
+	computed: {},
+	methods: {
+		async setToolOptions(new_value: number) {
+			// TODO: Each value-input widget (i.e. not a button) should map to a field in an options struct,
+			// and updating a widget should send the whole updated struct to the backend.
+			// Later, it could send a single-field update to the backend.
+			const { set_tool_options } = await wasm;
+			set_tool_options(this.$props.activeTool || "Select", { Shape: { shape_type: { Polygon: { vertices: new_value } } } });
+		},
+	},
+	data() {
+		return {
+			optionsMap: new Map([
```

Comment:
This is great for now but long-term we'd like the editor Rust code to specify the options bar widgets to render.

Reply:
Sounds good. I'd like to hear more about those plans, as #284 may be greatly simplified if we pushed this migration forward.